NoSuchMethodError: CuratorZookeeperClient.startAdvancedTracer when using ZookeeperBasedLockProvider with Spark 3.3

### Bug Description

**What happened:**
  When using Hudi 1.0.2 with Spark 3.3.4, enabling optimistic concurrency control with ZookeeperBasedLockProvider throws NoSuchMethodError:                                                                                                                                       
                                                       
```                                                                                                                                                                                                                                          
  java.lang.NoSuchMethodError: org.apache.curator.CuratorZookeeperClient.startAdvancedTracer(Ljava/lang/String;)Lorg/apache/curator/drivers/OperationTrace;                                                                                                                                      
      at org.apache.curator.framework.imps.ExistsBuilderImpl.pathInForegroundStandard(ExistsBuilderImpl.java:221)                                                                                                                                                                                
      at org.apache.hudi.client.transaction.lock.BaseZookeeperBasedLockProvider.createNodeIfNotExists(BaseZookeeperBasedLockProvider.java:115)
      at org.apache.hudi.client.transaction.lock.BaseZookeeperBasedLockProvider.<init>(BaseZookeeperBasedLockProvider.java:81)
      at org.apache.hudi.client.transaction.lock.ZookeeperBasedLockProvider.<init>(ZookeeperBasedLockProvider.java:44)
```

**What you expected:**
  Hudi 1.0.2 should work with Spark 3.3.4's bundled Curator version, or document the required Curator version for ZookeeperBasedLockProvider. The ZookeeperBasedLockProvider should either shade its Curator dependency or be compatible with the Curator version bundled in Spark.


**Steps to reproduce:**
  1. Set up environment: Spark 3.3.4 + Hudi 1.0.2 (bundled as --jars), Java 1.8.0_281, ZooKeeper 3.6.2 cluster
  2. Create a MOR Hudi table with optimistic concurrency control and ZK lock:
```sql  
CREATE TABLE test_table (
    col0 STRING, m79 STRING, ..., update_time TIMESTAMP
  ) USING HUDI
  TBLPROPERTIES (
    'type' = 'MERGE_ON_READ',
    'primaryKey' = 'col0',
    'preCombineField' = 'update_time',
    'hoodie.write.concurrency.mode' = 'OPTIMISTIC_CONCURRENCY_CONTROL',
    'hoodie.write.lock.provider' = 'org.apache.hudi.client.transaction.lock.ZookeeperBasedLockProvider',
    'hoodie.write.lock.zookeeper.url' = 'zk1:2181,zk2:2181,zk3:2181',
    'hoodie.write.lock.zookeeper.port' = '2181',
    'hoodie.write.lock.zookeeper.lock_key' = 'test_table',
    'hoodie.write.lock.zookeeper.base_path' = '/hudi/locks'
  );
```
  3. Execute a MERGE INTO or INSERT statement that triggers the lock provider. The operation fails immediately with HoodieException: Unable to instantiate class ZookeeperBasedLockProvider caused by NoSuchMethodError: CuratorZookeeperClient.startAdvancedTracer.



### Environment

**Hudi version:**
1.0.2
**Query engine:** (Spark/Flink/Trino etc)
Spark
**Relevant configs:**
    'hoodie.write.concurrency.mode' = 'OPTIMISTIC_CONCURRENCY_CONTROL',
    'hoodie.write.lock.provider' = 'org.apache.hudi.client.transaction.lock.ZookeeperBasedLockProvider',
    'hoodie.write.lock.zookeeper.url' = 'zk1:2181,zk2:2181,zk3:2181',
    'hoodie.write.lock.zookeeper.port' = '2181',
    'hoodie.write.lock.zookeeper.lock_key' = 'test_table',
    'hoodie.write.lock.zookeeper.base_path' = '/hudi/locks'

### Logs and Stack Trace

```
26/04/22 18:44:23 ERROR ExecuteStatement: Error operating ExecuteStatement: org.apache.hudi.exception.HoodieException: Unable to instantiate class org.apache.hudi.client.transaction.lock.ZookeeperBasedLockProvider
        at org.apache.hudi.common.util.ReflectionUtils.loadClass(ReflectionUtils.java:75)
        at org.apache.hudi.common.table.timeline.TimeGeneratorBase.getLockProvider(TimeGeneratorBase.java:93)
        at org.apache.hudi.common.table.timeline.TimeGeneratorBase.unlock(TimeGeneratorBase.java:117)
        at org.apache.hudi.common.table.timeline.SkewAdjustingTimeGenerator.generateTime(SkewAdjustingTimeGenerator.java:51)
        at org.apache.hudi.common.table.timeline.HoodieInstantTimeGenerator.lambda$createNewInstantTime$0(HoodieInstantTimeGenerator.java:78)
        at java.util.concurrent.atomic.AtomicReference.updateAndGet(AtomicReference.java:179)
        at org.apache.hudi.common.table.timeline.HoodieInstantTimeGenerator.createNewInstantTime(HoodieInstantTimeGenerator.java:75)
        at org.apache.hudi.common.table.timeline.TimelineUtils.generateInstantTime(TimelineUtils.java:542)
        at org.apache.hudi.common.table.timeline.TimelineUtils.generateInstantTime(TimelineUtils.java:530)
        at org.apache.hudi.client.BaseHoodieClient.createNewInstantTime(BaseHoodieClient.java:196)
        at org.apache.hudi.HoodieSparkSqlWriterInternal.writeInternal(HoodieSparkSqlWriter.scala:494)
        at org.apache.hudi.HoodieSparkSqlWriterInternal.$anonfun$write$1(HoodieSparkSqlWriter.scala:191)
        at org.apache.hudi.HoodieSparkSqlWriterInternal.write(HoodieSparkSqlWriter.scala:209)
        at org.apache.hudi.HoodieSparkSqlWriter$.write(HoodieSparkSqlWriter.scala:128)
        at org.apache.spark.sql.hudi.command.MergeIntoHoodieTableCommand.executeUpsert(MergeIntoHoodieTableCommand.scala:496)
        at org.apache.spark.sql.hudi.command.MergeIntoHoodieTableCommand.run(MergeIntoHoodieTableCommand.scala:278)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:75)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:73)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:84)
        at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:98)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:109)
        at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:95)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
        at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
        at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:98)
        at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:94)
        at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:584)
        at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:176)
        at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:584)
        at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:30)
        at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
        at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
        at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
        at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
        at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:560)
        at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:94)
        at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:81)
        at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:79)
        at org.apache.spark.sql.Dataset.<init>(Dataset.scala:219)
        at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:99)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
        at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:96)
        at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:622)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
        at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:617)
        at org.apache.kyuubi.engine.spark.operation.ExecuteStatement.$anonfun$executeStatement$1(ExecuteStatement.scala:86)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at org.apache.kyuubi.engine.spark.operation.SparkOperation.$anonfun$withLocalProperties$1(SparkOperation.scala:147)
        at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
        at org.apache.kyuubi.engine.spark.operation.SparkOperation.withLocalProperties(SparkOperation.scala:131)
        at org.apache.kyuubi.engine.spark.operation.ExecuteStatement.executeStatement(ExecuteStatement.scala:81)
        at org.apache.kyuubi.engine.spark.operation.ExecuteStatement$$anon$1.run(ExecuteStatement.scala:103)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.apache.hudi.common.util.ReflectionUtils.loadClass(ReflectionUtils.java:73)
        ... 57 more
Caused by: java.lang.NoSuchMethodError: org.apache.curator.CuratorZookeeperClient.startAdvancedTracer(Ljava/lang/String;)Lorg/apache/curator/drivers/OperationTrace;
        at org.apache.curator.framework.imps.ExistsBuilderImpl.pathInForegroundStandard(ExistsBuilderImpl.java:221)
        at org.apache.curator.framework.imps.ExistsBuilderImpl.pathInForeground(ExistsBuilderImpl.java:216)
        at org.apache.curator.framework.imps.ExistsBuilderImpl.forPath(ExistsBuilderImpl.java:175)
        at org.apache.curator.framework.imps.ExistsBuilderImpl.forPath(ExistsBuilderImpl.java:32)
        at org.apache.hudi.client.transaction.lock.BaseZookeeperBasedLockProvider.createNodeIfNotExists(BaseZookeeperBasedLockProvider.java:115)
        at org.apache.hudi.client.transaction.lock.BaseZookeeperBasedLockProvider.createPathIfNotExists(BaseZookeeperBasedLockProvider.java:105)
        at org.apache.hudi.client.transaction.lock.BaseZookeeperBasedLockProvider.<init>(BaseZookeeperBasedLockProvider.java:81)
        at org.apache.hudi.client.transaction.lock.ZookeeperBasedLockProvider.<init>(ZookeeperBasedLockProvider.java:44)
        ... 62 more
```

--- a/pom.xml
+++ b/pom.xml
@@ -492,6 +492,10 @@
               <shadedPattern>org.apache.hudi.com.fasterxml.jackson.module
               </shadedPattern>
             </relocation>
+            <relocation>
+              <pattern>org.apache.curator.</pattern>
+              <shadedPattern>org.apache.hudi.org.apache.curator.</shadedPattern>
+            </relocation>
           </relocations>
         </configuration>
       </plugin>

